### ISSUE TITLE
Always executing input with Enter?

### ISSUE THREAD

**Issue by Kwpolska** [ipython/ipython#10498]
In my opinion, this behavior is confusing (`|` is cursor):

```
>>> foo()| <enter>
foo runs
>>> foo(|) <enter>
a newline is added and editing continues
>>> foo(
|)
```

Could there be an option to make it work the old way (before 5.3.0)? I know I can work around it by using esc-enter (but I found out about this only by searching the issues and finding #10374 , so it’s undocumented!), however that’s not good enough for me (especially because I don’t need to it in my shell). I change parameters to functions more often than I write multi-line stuff in IPython.

via #9283 #10264 #10275 #10374

(IPython v6.0.0, Python 3.6.1, macOS Sierra)

**Comment by takluyver** [comment]
#10489 would make it do what you expect in this case.

**Comment by Carreau** [comment]
I like that as is and we should get it in or we won't try it.

@meeseeksdev merge

Thanks Thomas!

**Comment by carlsmith** [comment]
Is there any chance this could become 'officially' configurable please?

It'd be cool if users could pass a function in to the configuration that gets called every time Enter is pressed, that has a reference to the cli instance or whatever makes sense, so it could update state, adding newlines and indentation, moving the cursor position etc. then return a bool, indicating whether to execute or not.

Maybe there's a better API that can reuse PTK's transform machinery, and Pygments' tokens. I'm not sure, but would like a hook into the Enter key's action.

Having a few gists kicking about, that we can hack on as users, that cover the three or four ways this could work, would also stop us from debating which is best.

**Comment by takluyver** [comment]
> Is there any chance this could become 'officially' configurable please?

Sounds reasonable, #10500 would add a config option for it. The function you give it should work like [this one](https://github.com/ipython/ipython/blob/fcbde47323fac4809350e5ee21687027d42519f3/IPython/terminal/shortcuts.py#L97), which is the default implementation.

**Comment by Carreau** [comment]
closed by #10500
